source/reduce/reducer.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <sstream>
+
+#include "source/spirv_reducer_options.h"
+
+#include "reducer.h"
+#include "reduction_pass.h"
+
+namespace spvtools {
+namespace reduce {
+
+struct Reducer::Impl {
+  explicit Impl(spv_target_env env)
+      : target_env(env), made_progress_this_round(false) {}
+
+  const spv_target_env target_env;  // Target environment.
+  MessageConsumer consumer;         // Message consumer.
+  InterestingFunction is_interesting;
+  std::vector<std::unique_ptr<ReductionPass>> passes;
+  std::vector<std::unique_ptr<ReductionPass>>::iterator current_pass;
+  bool made_progress_this_round;
+
+  std::vector<uint32_t> ApplyReduction(const std::vector<uint32_t>& binary);
+  bool Finished();
+};
+
+Reducer::Reducer(spv_target_env env) : impl_(MakeUnique<Impl>(env)) {}
+
+Reducer::~Reducer() = default;
+
+void Reducer::SetMessageConsumer(MessageConsumer c) {
+  for (auto& pass : impl_->passes) {
+    pass->SetMessageConsumer(c);
+  }
+  impl_->consumer = std::move(c);
+}
+
+void Reducer::SetInterestingFunction(
+    Reducer::InterestingFunction interestingFunction) {
+  impl_->is_interesting = std::move(interestingFunction);
+}
+
+bool Reducer::Run(std::vector<uint32_t>&& binary_in,
+                  std::vector<uint32_t>& binary_out,
+                  spv_const_reducer_options options) const {
+  impl_->current_pass = impl_->passes.begin();
+
+  uint32_t current_step = 0;
+
+  std::vector<uint32_t> current = std::move(binary_in);
+
+  // Initial state should be interesting.
+  assert(impl_->is_interesting(current));
+
+  for (;;) {
```

Comment:
Instead of the following assert, why not enforce bounds on the loop?

Reply:
Good plan - done.